api/src/main/java/org/openmrs/module/registrationcore/api/impl/PatientCreationListener.java

```diff
@@ -0,0 +1,102 @@
+package org.openmrs.module.registrationcore.api.impl;
+
+import org.openmrs.Patient;
+import org.openmrs.PatientIdentifier;
+import org.openmrs.User;
+import org.openmrs.api.UserService;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.context.Daemon;
+import org.openmrs.event.Event;
+import org.openmrs.module.DaemonToken;
+import org.openmrs.module.registrationcore.RegistrationCoreConstants;
+import org.openmrs.module.registrationcore.api.mpi.common.MpiProperties;
+
+import javax.jms.Message;
+
+/**
+ * This class listen for patient creation event.
+ * If MPI is enabled it perform export patient to MPI,
+ * and perform update for local patient with new one patient identifier.
+ */
+public class PatientCreationListener extends PatientActionListener {
+
+    private MpiProperties mpiProperties;
+
+    private IdentifierBuilder identifierBuilder;
+
+    private UserService userService;
+
+    private DaemonToken daemonToken;
+
+    public void setMpiProperties(MpiProperties mpiProperties) {
+        this.mpiProperties = mpiProperties;
+    }
+
+    public void setIdentifierBuilder(IdentifierBuilder identifierBuilder) {
+        this.identifierBuilder = identifierBuilder;
+    }
+
+    public void setUserService(UserService userService) {
+        this.userService = userService;
+    }
+
+    public void setDaemonToken(DaemonToken daemonToken) {
+        this.daemonToken = daemonToken;
+    }
+
+    /**
+     * Subscribes for patient creation event.
+     */
+    public void init() {
+        Event.subscribe(RegistrationCoreConstants.PATIENT_REGISTRATION_EVENT_TOPIC_NAME, this);
+    }
+
+    /**
+     * Export patient to MPI server & update local
+     * patient with new one patient identifier.
+     *
+     * @param message message with properties.
+     */
+    @Override
+    public void performMpiAction(Message message) {
+        Patient patient = extractPatient(message);
+
+        String personId = pushPatientToMpiServer(patient);
+
+        updatePatient(patient, message, personId);
+    }
+
+    private String pushPatientToMpiServer(Patient patient) {
+        return coreProperties.getMpiProvider().exportPatient(patient);
+    }
+
+    private void updatePatient(final Patient patient, final Message message, final String personId) {
+        Daemon.runInDaemonThread(new Runnable() {
+            @Override
+            public void run() {
+                Context.openSession();
```

Comment:
You should not need to do Context.open/closeSession, because the Daemon takes care of that, right?